README.md
simple-sentinel [![Build Status](https://travis-ci.org/Vungle/simple-sentinel.svg?branch=master)](https://travis-ci.org/Vungle/simple-sentinel)
===============

A stable, and easy to use redis-sentinel client for Node.js. Features include:

 - Simple, event-based API.
 - Randomized connections to sentinels, and automatic reconnection on sentinel failure.
 - Uses Sentinel's Pub/Sub commands for real-time config updates during failovers.
 - Uses Sentinel info to help dodge outages.

### Installing

```sh
npm install --save simple-sentinel
```

### Sample Code

```javascript
var RedisSentinel = require('simple-sentinel');

// Put all sentinel servers in here. A random one will be used:
var sentinels = [
  { host: '10.2.3.41', port: 26379 },
  { host: '10.2.3.42', port: 26379 },
  { host: '10.2.3.43', port: 26379 },
  { host: '10.2.3.44', port: 26379 },
  { host: '10.2.3.45', port: 26379 }
];

var options = {
  // Omit this to track all servers that the sentinel knows about:
  watchedNames: ["replica_a", "replica_b", "replica_d"]
};

// Create a sentinel object to track the given Replicas:
var sentinel = new RedisSentinel(sentinels, options);

// Keep track of the Masters here:
var masters = {
  replica_a: null,
  replica_b: null,
  replica_d: null
};

// Listen for connection info changes:
sentinel.on('change', function (name, replica) {
  console.log("Just got connection info for Replica:", name, replica.toString());
  masters[name] = replica.connectMaster();
});
```

### API

#### RedisSentinel

This object is used to connect to a redis-sentinel cluster. It is an EventEmitter, and will notify you about outages or failovers as they happen.

##### Constructor: (sentinels, config)

Will open the connection to one of the sentinels in the 'sentinels' array.

- `sentinels` is an Array of objects, each with connection info for a single redis-sentinel. These objects should contain:
    - `host` (**String**) is the hostname to connect to.
    - `port` (**Number**) is the port to connect to. Default: 26379.

- `config` is an optional object used to store other configuration details. If omitted (or otherwise falsey) then only default values will be used. This object can contain:
    - `commandTimeout` (**Number**) is the maximum number of milliseconds that we'll wait for a command on this sentinel to return. Default is 1500.
    - `createClient` (**Function**) is the function that will create the `RedisClient`s that are returned to you. By default, this library will try to require() the [node-redis](https://github.com/mranney/node_redis) library from the scope of your project automagically, and use our own version if yours is unavailable. If you want to do something more advanced, then you can use this argument to override the default. The function will be called with the arguments: `(port, host, options)`.
    - `customLogger` (**Function**) is the function used when logging messages. This function will only be called if `debugLogging` is set to true. Is called with a single string argument: `(message)`. Default: `console.log`.
    - `debugLogging` (**Boolean**) should be true if you want to see the noisy debug log. Default is `false`.
    - `outageRetryTimeout` (**Number**) is the number of milliseconds before trying again if ALL sentinels are down. If this number is negative, then we will simply emit an error instead of retrying. Default is 5000.
    - `randomizeSentinels` (**Boolean**) indicates whether the sentinels list should be shuffled before attempting connections. Default is true.
    - `redisOptions` (**Object**) is the object that is passed to createClient as options. Default is `{}`.
    - `refreshTimeout` (**Number**) is the number of milliseconds between attempts to fetch configurations from redis. Normally, we use events emitted to us through sentinel's pub/sub mechanism, but we always have this as a fallback should a message not make it to us. Default is 60000.
    - `timeout` (**Number**) is the connect timeout in milliseconds for connecting to a sentinel server. Default is 500.
    - `watchedNames` (**Array**) is an Array of the String names of Replicas to watch. Default is to watch everything that a sentinel is currently watching.

##### Event: "change"

Emitted whenever the connection information for a Replica Set has changed. Use this event to build RedisClient's for all of the servers that you're keeping track of. This will also be emitted during startup, so you can connect to Redis quickly.

Has arguments: `(name, replica)`
- `name` (**String**) is name of a watched Replica.
- `replica` (**RedisReplica**) is a `RedisReplica` object. Use this object to connect to masters or slaves. For the full API, [see below](#redisreplica).

##### Event: "error"

Emitted when there was an internal error, or if all redis-sentinels are down and retries have been disabled.

Has arguments: `(error)`
- `error` is an Error object. Big surprise. :)

##### Event: "event"

Emitted when an important event has happened. These are simply events that were passed to us through pub/sub in sentinel. Don't rely on these: some events may be omitted or duplicated in certain cases (such as when switching from one sentinel to another.) This is mostly useful for logging.

Has arguments: `(event_type, event_message)`
- `event_type` (**String**) is the event type, like `"+odown"`.
- `event_message` (**String**) describes what happened, like `"master main 127.0.0.1 9001 #quorum 4/3"`.


#### RedisReplica

This object stores the connection information to all object in a Replica Set, and gives you a bunch of useful methods for connecting to them with [node-redis](https://github.com/mranney/node_redis). It is provided in the `change` events.

##### Function: connectMaster()

Will create a connection to the master server in the Replica Set. Returns a brand-new `RedisClient`. Will be `null` if the server is currently objectively down, and there is no way to connect to it. (Uh-oh.)

##### Function: connectSlave()

Will create a connection to a random available slave in the Replica Set. A slave is unavailable if the Sentinel describes this server as being subjectively down. Returns a brand-new `RedisClient`. Will be `null` if no slaves are currently available.

##### Function connectAllSlaves()

Will create connections to all available slaves in a Replica Set. A slave is unavailable if the Sentinel describes this server as being subjectively down. Returns an Array of brand-new `RedisClient`s. Useful if you want to implement any round-robin load balancing, or what-have-you. This Array will be empty if no slaves are currently connectable.

##### Function: getMasterConfig()

Will fetch the connection information for the current master in the Replica Set. If the master is available, then returns an Object with the keys:
- `host` (**String**) is the hostname of the master. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

Returns `null` when master is unavailable.

##### Function: getSlaveConfig()

Will fetch the connection information for a random slave in the Replica Set. If a slave is available, then returns an Object with the keys:
- `host` (**String**) is the hostname of the slave. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

Returns `null` when all slaves are down.

##### Function: getAllSlaveConfigs()

Will fetch the connection information for all available slaves in the Replica Set. Returns an array of Objects, where each object has keys:
- `host` (**String**) is the hostname of the slave. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

### License

MIT


CHANGELOG.md
## Changelog

### 0.2.0 (*2015 Jul 2*)
 - Added API to get connection info without connecting.
 - Fixed a minor bug with error handling.

### 0.1.2 (*2014 Dec 14*)
 - Fixed write-to-global bug in the logging logic.
 - Added ability to use custom logging functions.

### 0.1.1 (*2014 Nov 26*)
 - NPM release.
 - Doc updates.
 - Sentinels array is now deduped.
 - Test fixes.

### 0.1.0 (*2014 Nov 18*)
 - Initial release. (Not on NPM though...)
 - Random sentinel connection to a sentinel group.
 - Master / Slave detection for replica sets.
 - Push notifications from sentinel on configuration changes.
 - Up / Down detection for RedisClient connections.
 - Can connect to masters, a random slave, or all slaves simultaneously.


tests/util_test.js
var util = require('../src/util')
  , mockRandom = require('./util/mockRandom')
  , fs = require('fs')
  , path = require('path')
  , temp = require('temp').track()
  , expect = require('expect');

describe('Util', function () {

  it("extends the Node.js util lib", function () {
    expect(util.format).toBeA(Function);
    expect(util.inherits).toBeA(Function);
  });

  describe("when require()-ing from a parent", function () {

    beforeEach(function () {
      // Blank the util cache reference:
      delete require.cache[require.resolve('../src/util')];
    });

    afterEach(function () {
      // Restore the util cache to a non-temp require chain:
      delete require.cache[require.resolve('../src/util')];
      util = require('../src/util');
    });

    it("gets the right module", function () {

      // This test is tricky. We need to create a temp directory with a sub
      // directory. We then test the ability of one file to use the parent
      // require function with a relative path, and make sure that the path
      // correctly evaluates from the parent directory.
      //
      // We use sync functions here, because this is a test and I'm not insane.

      var util_dir         = path.join(__dirname, "../src/util");
      var util_dir_escaped = util_dir.replace(/'/g, "\\'");
      var temp_dir         = temp.mkdirSync("sentinel_require_test");
      var temp_subdir      = path.join(temp_dir, "dir");

      var test_start_path = path.join(temp_dir, "run_test.js");

      fs.mkdirSync(temp_subdir);

      fs.writeFileSync(test_start_path,                  "module.exports=require('./dir');");
      fs.writeFileSync(path.join(temp_dir, "dep.js"),    "module.exports='root_dir';");

      fs.writeFileSync(path.join(temp_subdir, "dep.js"), "module.exports='sub_dir';");
      fs.writeFileSync(path.join(temp_subdir, "index.js"),
        "var util = require('"+util_dir_escaped+"');\n" +
        "module.exports = {\n" +
        "  ours:   require('./dep'),\n" +
        "  parent: util.parentRequire('./dep')\n" +
        "};\n"
      );

      // Require the bottom object, and see what it comes up with:
      var result = require(test_start_path);
      expect(result.ours).toBe("sub_dir");
      expect(result.parent).toBe("root_dir");
    });

    it("gets null on not found", function () {
      // This module BETTER not exist:
      var out = util.parentRequire("fdsjafdsajkfldsjaifodlkfsdjalkf-" + Date.now());
      expect(out).toBe(null);
    });
  });

  describe("when making a logger", function () {

    // Swap out the console.log, so we don't get annoying output while testing:

    var last_log = null;

    function _customLogger(msg) {
      last_log = msg;
    }

    afterEach(function () {
      last_log = null;
    });

    it("will create a muted log", function () {
      var logger = util.buildLogger("Derp");
      logger("Sup");
      expect(last_log).toBe(null);
    });

    it("can unmute a log", function () {
      var logger = util.buildLogger("Derp");
      logger.configure({debugLogging: true, customLogger: _customLogger});
      logger("Sup");
      expect(last_log).toBe("Derp: Sup");
    });

    it("will accept formatting stuff", function () {
      var logger = util.buildLogger("Derp");
      logger.configure({debugLogging: true, customLogger: _customLogger});
      logger("Hello %s!!!%d!", "World", 1);
      expect(last_log).toBe("Derp: Hello World!!!1!");
    });

    describe("with errors", function () {
      it("will show stack traces if available", function () {
        var logger = util.buildLogger("Derp");
        logger.configure({debugLogging: true, customLogger: _customLogger});
        logger("Error encountered:", new Error("I AM AN ERROR"));
        expect(last_log).toMatch(/^Derp: Error encountered: /);
        expect(last_log).toMatch(/I AM AN ERROR/);
        expect(last_log).toMatch(/util_test\.js/);
      });

      it("will stringify an error if no stack trace", function () {
        var logger = util.buildLogger("Derp");
        var err = new Error("I AM AN ERROR");
        delete err.stack;
        logger.configure({debugLogging: true, customLogger: _customLogger});
        logger("Error encountered:", err);
        expect(last_log)
          .toMatch(/^Derp: Error encountered: /)
          .toMatch(/I AM AN ERROR/);
      });
    });
  });

  describe("when shuffling an array", function () {

    mockRandom.installHooks();

    it("both edits in place and returns the shuffled array", function () {
      var a = [1,2,3,4,5,6,7,8,9];
      var res = util.shuffleArray(a);

      expect(a)
        .toExist()
        .toBeAn(Array);

      expect(res)
        .toExist()
        .toBeAn(Array);

      for (var idx=0; idx<a.length; idx++) {
        expect(a[idx]).toBe(res[idx], "Shuffled arrays differ at index " + idx);
      }
    });

    it("gives an ok distribution", function () {
      this.timeout(10000);
      var array;
      var ours = {};
      for(var i=0; i<48000; i++) {
        array = ["1", "2", "3", "4"];
        util.shuffleArray(array);
        var key = array.join("");
        ours[key] = (ours[key] || 0) + 1;
      }

      Object.keys(ours).forEach(function (perm) {
        // Biased solutions cause greater variations in the counts.
        // Good ones will split evenly between the 24 possible permutations,
        // so 2000 each. Bad shuffles will accidentally give some orderings
        // preference, so we use 1700 and 2300 as cutoffs.
        expect(ours[perm])
          .toBeGreaterThan(1700, "Detected a shuffle bias for: " + perm)
          .toBeLessThan   (2300, "Detected a shuffle bias for: " + perm);
      });
    });
  });
});


tests/util/mockRandom.js
var rand = require('random-seed');


// Random tests suck, so we swap out Math.random for a good PRNG with a fixed seed,
// so that we can have all the fun of a random() function with a high period, and
// yet still be consistent and repeatable.


var SAVED_SEED = 382948;

module.exports.installHooks = function () {

  // Swap out Math.random before tests:
  var saved_random = null;
  var rand_engine = null;

  before(function initRNG() {
    rand_engine = rand.create(SAVED_SEED);
  });

  after(function releaseRNG() {
    rand_engine.done();
    rand_engine = null;
  });

  beforeEach(function swapOut() {
    saved_random = Math.random;

    Math.random = function () {
      var out = rand_engine.random();
      return out;
    };
  });

  // Repair it after each test:
  afterEach(function repair() {
    Math.random = saved_random;
    saved_random = null;
  });
};


tests/util/mockSentinel.js
var net = require('net');
var boris = require("boris");
var util = require('../../src/util');
var events = require('events');


// It's hard to test sentinel property, especially when we want to support environments
// like TravisCI, where they have only very simple redis support, and no redis-sentinel
// support.
//
// So we have this. It is a simple TCP server that responds exactly as you configure it
// to, and speaks the Redis protocol.


var INFO_STRINGS = {
  normal:
    "# Server\r\nredis_version:2.8.11\r\nredis_git_sha1:00000000\r\nredis_git_dirty:0" +
    "\r\nredis_build_id:cd43e547b41f72f\r\nredis_mode:standalone\r\nos:Darwin 13.4.0 " +
    "x86_64\r\narch_bits:64\r\nmultiplexing_api:kqueue\r\ngcc_version:4.2.1\r\nproces" +
    "s_id:56422\r\nrun_id:8c786b8dd9e86ff83d360ac0c34a7632afbbff66\r\ntcp_port:6379\r" +
    "\nuptime_in_seconds:25781\r\nuptime_in_days:0\r\nhz:10\r\nlru_clock:6747792\r\nc" +
    "onfig_file:/usr/local/etc/redis.conf\r\n\r\n# Clients\r\nconnected_clients:9\r\n" +
    "client_longest_output_list:0\r\nclient_biggest_input_buf:0\r\nblocked_clients:0" +
    "\r\n\r\n# Memory\r\nused_memory:1145552\r\nused_memory_human:1.09M\r\nused_memory" +
    "_rss:2015232\r\nused_memory_peak:1499792\r\nused_memory_peak_human:1.43M\r\nused" +
    "_memory_lua:33792\r\nmem_fragmentation_ratio:1.76\r\nmem_allocator:libc\r\n\r\n#" +
    " Persistence\r\nloading:0\r\nrdb_changes_since_last_save:0\r\nrdb_bgsave_in_prog" +
    " ress:0\r\nrdb_last_save_time:1416014109\r\nrdb_last_bgsave_status:ok\r\nrdb_last" +
    "_bgsave_time_sec:0\r\nrdb_current_bgsave_time_sec:-1\r\naof_enabled:0\r\naof_rew" +
    "rite_in_progress:0\r\naof_rewrite_scheduled:0\r\naof_last_rewrite_time_sec:-1\r" +
    "\naof_current_rewrite_time_sec:-1\r\naof_last_bgrewrite_status:ok\r\naof_last_wri" +
    "te_status:ok\r\n\r\n# Stats\r\ntotal_connections_received:33\r\ntotal_commands_p" +
    "rocessed:343\r\ninstantaneous_ops_per_sec:0\r\nrejected_connections:0\r\nsync_fu" +
    "ll:0\r\nsync_partial_ok:0\r\nsync_partial_err:0\r\nexpired_keys:0\r\nevicted_key" +
    "s:0\r\nkeyspace_hits:41\r\nkeyspace_misses:126\r\npubsub_channels:0\r\npubsub_pa" +
    "tterns:0\r\nlatest_fork_usec:233\r\n\r\n# Replication\r\nrole:master\r\nconnecte" +
    "d_slaves:0\r\nmaster_repl_offset:0\r\nrepl_backlog_active:0\r\nrepl_backlog_size" +
    ":1048576\r\nrepl_backlog_first_byte_offset:0\r\nrepl_backlog_histlen:0\r\n\r\n# " +
    "CPU\r\nused_cpu_sys:1.21\r\nused_cpu_user:1.15\r\nused_cpu_sys_children:0.00\r\n" +
    "used_cpu_user_children:0.00\r\n\r\n# Keyspace\r\ndb0:keys=1,expires=0,avg_ttl=0" +
    "\r\ndb1:keys=11,expires=5,avg_ttl=101643976\r\ndb3:keys=3,expires=0,avg_ttl=0\r" +
    "\ndb4:keys=7,expires=3,avg_ttl=926438415\r\ndb9:keys=1,expires=1,avg_ttl=665737" +
    "58\r\n",

  sentinel:
    "# Server\r\nredis_version:2.8.11\r\nredis_git_sha1:00000000\r\nredis_git_dirty:0" +
    "\r\nredis_build_id:cd43e547b41f72f\r\nredis_mode:sentinel\r\nos:Darwin 13.4.0 x8" +
    "6_64\r\narch_bits:64\r\nmultiplexing_api:kqueue\r\ngcc_version:4.2.1\r\nprocess_" +
    "id:48377\r\nrun_id:8397fda691ee9289b623c3904e21301ac4ef82ae\r\ntcp_port:9501\r\n" +
    "uptime_in_seconds:116301\r\nuptime_in_days:1\r\nhz:16\r\nlru_clock:6748294\r\nco" +
    "nfig_file:/Users/lolusername/sentinel-test/a/sentinel.conf\r\n\r\n# Sentinel\r\n" +
    "sentinel_masters:3\r\nsentinel_tilt:0\r\nsentinel_running_scripts:0\r\nsentinel_" +
    "scripts_queue_length:0\r\nmaster0:name=cache,status=ok,address=127.0.0.1:9203,sl" +
    "aves=2,sentinels=5\r\nmaster1:name=main,status=ok,address=127.0.0.1:9001,slaves=" +
    "1,sentinels=5\r\nmaster2:name=shard,status=ok,address=127.0.0.1:9102,slaves=1,se" +
    "ntinels=5\r\n"
};


function MockSentinel() {
  var sentinel = this;

  this.info = null;
  this.masters = {};
  this.slaves = {};
  this.was_started = false;
  this.stop_responding = false;

  this.server = net.createServer(function (socket) {

    sentinel.emit('connection');

    var parser = boris({return_buffers: false});
    var subscribed_on = [];

    parser.on('match', function (err, cmd) {
      if (err) {
        socket.destroy();
        throw err;
      }

      var arr;

      if (sentinel.stop_responding) { return; }

      switch(cmd[0] && cmd[0].toLowerCase()) {
        case 'info':
          if (sentinel.info !== null) {
            socket.write("$" + sentinel.info.length + "\r\n" + sentinel.info + "\r\n");
          }
          break;

        case 'subscribe':
          var new_channels = cmd.slice(1);
          subscribed_on = subscribed_on.concat(new_channels);
          var response = new_channels
            .map(function (ch, idx) {
              return "*3\r\n$9\r\nsubscribe\r\n$" + ch.length + "\r\n" + ch + "\r\n:" + (idx+1) + "\r\n";
            })
            .join("");
          socket.write(response);
          break;

        case 'sentinel':
          switch(cmd[1] && cmd[1].toLowerCase()) {
            case 'masters':
              // Build masters arrays:
              arr = Object.keys(sentinel.masters)
                .map(function (name) {
                  var obj = sentinel.masters[name];
                  var row = [];
                  row.push("name", name);
                  row.push("ip", obj.ip);
                  row.push("port", obj.port);
                  row.push("flags", "master" + (obj.isDown ? ",o_down" : ""));
                  return row;
                });
              socket.write(serializeArray(arr));
              break;

            case 'slaves':
              // Build masters arrays:
              arr = (sentinel.slaves[cmd[2]] || [])
                .map(function (obj) {
                  var row = [];
                  row.push("name", obj.ip + ":" + obj.port);
                  row.push("ip", obj.ip);
                  row.push("port", obj.port);
                  row.push("flags", "slave" + (obj.isDown ? ",s_down" : ""));
                  return row;
                });
              socket.write(serializeArray(arr));
              break;

            default:
              socket.destroy();
              throw new Error("Unsupported SENTINEL command:" + String(cmd));
          }
          break;

        default:
          // Unknown command:
          socket.destroy();
          throw new Error("Unsupported command: " + String(cmd));
      }
    });

    sentinel.on('event', function (name, msg) {
      if (subscribed_on.indexOf(name) >= 0) {
        // Send the message:
        var output = "*3\r\n$7\r\nmessage\r\n$" + name.length + "\r\n" + name + "\r\n$" + msg.length + "\r\n" + msg + "\r\n";
        socket.write(output);
      }
    });

    sentinel.on('suicide', function () {
      socket.destroy();
    });

    socket.on('data', function (d) {
      parser.parse(d);
    });
  });
}

util.inherits(MockSentinel, events.EventEmitter);

MockSentinel.prototype.setInfo = function (type) {
  if (!INFO_STRINGS.hasOwnProperty(type)) { throw new Error("Invalid type"); }
  this.info = INFO_STRINGS[type];
  return this;
};

MockSentinel.prototype.addMaster = function (name, obj) {
  if (!name || !obj || typeof name !== 'string' || typeof obj !== 'object') {
    throw new Error("Wrong args passed to addMaster");
  }
  // obj is like: {ip: "...", port: 6379, isDown: true}
  this.masters[name] = obj;
  return this;
};

MockSentinel.prototype.addSlave = function (name, obj) {
  if (!name || !obj || typeof name !== 'string' || typeof obj !== 'object') {
    throw new Error("Wrong args passed to addSlave");
  }
  // obj is like: {ip: "...", port: 6379, isDown: true}
  this.slaves[name] = this.slaves[name] || [];
  this.slaves[name].push(obj);
  return this;
};

MockSentinel.prototype.stopResponding = function () {
  this.stop_responding = true;
  return this;
};

MockSentinel.prototype.start = function (cb) {
  var server = this.server;
  this.was_started = true;
  // Listen on random port, but only respond to localhost:
  server.listen(0, '127.0.0.1', function () {
    var addr = server.address();
    return cb(null, addr.port);
  });
};

MockSentinel.prototype.kill = function (cb) {
  if (!this.was_started) { return cb && cb(); }
  this.emit('suicide');
  this.was_started = false;
  this.server.close(cb);
};

MockSentinel.prototype.sendEvent = function (name, msg) {
  this.emit('event', name, msg);
};


function serializeArray(arr) {
  var out = "*" + arr.length + "\r\n";

  arr.forEach(function (item) {
    if (Array.isArray(item)) {
      out += serializeArray(item);
    } else {
      var str = String(item);
      out += "$" + str.length + "\r\n" + str + "\r\n";
    }
  });

  return out;
}

module.exports = MockSentinel;


src/util.js
var node_util = require('util')
  , _         = require('lodash');


// Make a copy of Node's util:
var util    = _.extend({}, node_util);


/**
 * Will try to require something from a parent module.
 *
 * @param  {String} module_name The name of the module to require. Like the args to a normal require.
 * @return {Object}             The require()-ed module, or null if we couldn't find it at any level.
 */
function parentRequire(module_name) {

  // One layer up is our own index.js. Two layers up is the thing that required this lib:
  var m = module.parent && module.parent.parent;

  while ( m ) {
    try {
      return m.require(module_name);
    } catch (ex) {
      m = m.parent;
    }
  }

  return null;
}


/**
 * Will shuffle an array in-place with the Fisher-Yates algorithm.
 *
 * @param  {Array} arr The array to shuffle
 * @return {Array}     The array, for convenience.
 */
function shuffleArray(arr) {
  var i, rnd, tmp
    , len = arr.length;

  for (i = 0 ; i < len-1; i++) {
    rnd = i + Math.floor(Math.random() * (len - i));
    tmp = arr[i];
    arr[i] = arr[rnd];
    arr[rnd] = tmp;
  }

  return arr;
}


/**
 * Will dispatch a Redis command that can time out.
 *
 * @param  {RedisClient} client  The redis client to use.
 * @param  {Number}      timeout The number of milliseconds to wait.
 * @param  {String}      cmd     The command name.
 * @param  {String}      opts    An array of options. Optional.
 * @param  {Function}    cb      Called with (err, results)
 */
function timedCommand(client, timeout, cmd, opts, cb) {
  if (!cb) {
    cb = opts;
    opts = [];
  }

  cb = _.once(cb);

  var cmd_str = util.format("%s %s", cmd, opts.join(" "));
  var timeout_err = new Error("Command timed out: " + cmd_str);
  var cmd_timeout = setTimeout( cb.bind(null, timeout_err), timeout );

  client.send_command(cmd, opts, function () {
    clearTimeout(cmd_timeout);
    cb.apply(null, arguments);
  });
}


/**
 * Will build a simple logger. Will write to stdout IFF we were told to in the user configs.
 * Arguments are the same format as for node util's format function, with the exception that
 * all instances of Error with a 'stack' property are evaluated as that, and as a string
 * otherwise.
 */
function buildLogger(namespace) {
  namespace = String(namespace) + ": ";

  var enabled   = false
    , logger_fn = console.log.bind(console);

  var out = function _log() {
    if ( ! enabled ) { return; }

    var args = [].slice.call(arguments, 0);
    var i, len = args.length;
    for (i=0; i<len; i++) {
      var arg = args[i];
      if (arg instanceof Error) {
        args[i] = arg.stack || String(arg);
      }
    }
    var str = util.format.apply(util, args);
    logger_fn(namespace + str);
  };

  out.configure = function configure(options) {
    enabled = !! options.debugLogging;
    if (typeof options.customLogger === 'function') {
      logger_fn = options.customLogger;
    }
  };

  return out;
}


// Now to put our own stuff in there:
util._             = _;
util.async         = require('async');
util.parentRequire = parentRequire;
util.shuffleArray  = shuffleArray;
util.timedCommand  = timedCommand;
util.buildLogger   = buildLogger;


module.exports = util;


src/redisSentinel.js
var events = require('events')
  , redis = require('redis')
  , RedisReplica = require('./redisReplica')
  , RedisConfigFetcher = require('./redisConfigFetcher')
  , RedisWatcher = require('./redisWatcher')
  , util = require('./util');


// This is passed into the RedisReplica, so that the client redis objects are their own:
var client_redis = null;

// Uncomment this to debug the redis protocol stuffs:
// redis.debug_mode = true;


function RedisSentinel(sentinels, options) {
  if (!(this instanceof RedisSentinel)) {
    return new RedisSentinel(sentinels, options);
  }

  // Add in the things that will be asynchronously populated later:
  this.fetcher = null;
  this.watcher = null;
  this.replicas = {};

  // Keep track of death:
  this.terminated = false;

  // Simple validation:
  RedisSentinel._validateSentinelList(sentinels);

  // Extend the default values with the custom parameters:
  var defaults = {
    commandTimeout:     1500,
    createClient:       client_redis && client_redis.createClient || redis.createClient,
    customLogger:       null,
    debugLogging:       false,
    outageRetryTimeout: 5000,
    randomizeSentinels: true,
    redisOptions:       {},
    refreshTimeout:     60000,
    timeout:            500,
    watchedNames:       null,
  };

  this.options = util._.extend(defaults, options);

  // Make sure that the outside can't inject a test client:
  delete this.options._testClient;

  this._log = util.buildLogger("RedisSentinel");
  this._log.configure(this.options);

  this._log("Initialization started. Has client redis:", !!client_redis);

  // Store connection info for all the sentinels:
  this.sentinels = sentinels.map(function (conf) {
    var host = String(conf.host)
      , port = parseInt(conf.port, 10) || 26379;

    return { host: host, port: port };
  });

  this._dedupeSentinels();

  // Start this sucker on its way:
  this._connectSentinel();
}

// Inherit from EventEmitter, so we can emit stuff:
util.inherits(RedisSentinel, events.EventEmitter);


/**
 * Will do basic validation on a sentinel list. Throws if there was a problem.
 *
 * @param  {Array} sentinels The sentinel list, from the user.
 */
RedisSentinel._validateSentinelList = function _validateSentinelList(sentinels) {

  // First, check type:
  if (!sentinels || !Array.isArray(sentinels)) {
    throw new TypeError("First argument needs to be an array of sentinel config objects");
  }

  // Check array length:
  if (!sentinels.length) {
    throw new TypeError("Sentinels array shouldn't be empty");
  }

  // Now go through each item in the list, and make sure that there's the necessary info:
  sentinels.forEach(function (item, idx) {
    if (!item || typeof item !== 'object') {
      throw new TypeError("Item #" + idx + " in sentinels array isn't an object");
    }
    if (!item.host || typeof item.host !== 'string') {
      throw new Error("Item #" + idx + " in sentinels array doesn't have a correct host property");
    }
  });
};


/**
 * Will try to connect to each item in the sentinels array, in order, until it is
 * successful.
 */
RedisSentinel.prototype._connectSentinel = function _connectSentinel() {
  var sentinel = this;

  this._log("Starting sentinel connection...");

  // Make sure we are iterating over the sentinels in a random order:
  if (this.options.randomizeSentinels) {
    util.shuffleArray(this.sentinels);
  }

  // Call this when shit fucks up:
  var _abortAndReconnect = util._.once(sentinel._handleErrorAndReconnect.bind(sentinel));

  // Try them all!
  util.async.forEachSeries(
    this.sentinels,
    function withEachSentinel(conf, next) {

      var fetcher = sentinel.fetcher = new RedisConfigFetcher(conf.host, conf.port, sentinel.options)
        .once('error', _onError)
        .once('connected', _onConnect);

      // Errors at this stage mean that we will move to the next sentinel:
      function _onError() {
        fetcher.removeListener('connected', _onConnect);
        if (sentinel.terminated) { return; }
        next();
      }

      // Else, cool. Re-hook this thing up in a more stable way, and ask it for a config:
      function _onConnect() {
        fetcher.removeListener('error', _onError);
        if (sentinel.terminated) { return; }

        // Hook up the event handlers for the long run:
        fetcher
          .on('error', _abortAndReconnect)
          .on('config', _onGetReplInfo);

        // Now, create a watcher to poke this sucker along:
        sentinel.watcher = new RedisWatcher(conf.host, conf.port, sentinel.options)
          .on('error', _abortAndReconnect)
          .on('event', _passAlongEvent)
          .on('refresh', fetcher.updateConfigs.bind(fetcher));

        function _passAlongEvent(channel, msg) {
          if (sentinel.terminated) { return; }
          sentinel.emit('event', channel, msg);
        }

        function _onGetReplInfo(name, master, slaves) {
          if (sentinel.terminated) { return; }

          // Create a new RedisReplica if we've never seen this name before:
          if (!sentinel.replicas.hasOwnProperty(name)) {
            sentinel.replicas[name] = new RedisReplica(name, sentinel.options.createClient, sentinel.options.redisOptions);
          }

          // Load the master / slave configs into the struct:
          var repl           = sentinel.replicas[name]
            , master_changed = repl._loadMasterConfig(master)
            , slaves_changed = repl._loadSlaveConfigs(slaves)
            , has_changed    = master_changed || slaves_changed;

          sentinel._log("Repl configuration set:", repl.toString());

          // Emit the repl if things are different:
          if (has_changed) {
            sentinel._log("Repl", name, "has changed");
            sentinel.emit('change', name, repl);
          }
        }
      }
    },
    function (err) {
      if (sentinel.terminated) { return; }
      if (err) {
        sentinel.emit("error", err);
        return;
      }

      // We made it through all endpoints. What do??
      if (sentinel.options.outageRetryTimeout < 0) {
        // Stop, and emit error:
        sentinel.emit("error", new Error("Could not connect to a sentinel. *<:'O("));
        return;

      } else {
        // Loop around and try again in a few seconds:
        sentinel._log("All sentinels down. Pausing before retry...");
        setTimeout(sentinel._connectSentinel.bind(sentinel), sentinel.options.outageRetryTimeout);
      }
    }
  );
};


RedisSentinel.prototype._handleErrorAndReconnect = function _handleErrorAndReconnect(err) {
  if (this.terminated) { return; }

  // Log the error:
  this._log("Error encountered:", err);

  // Kill our redis clients:
  if (this.fetcher) { this.fetcher.kill(); }
  if (this.watcher) { this.watcher.kill(); }
  this.fetcher = this.watcher = null;

  // Trigger a re-connect:
  this._connectSentinel();
};


RedisSentinel.prototype._dedupeSentinels = function _dedupeSentinels() {
  // Dedupe the list based on "host(lowercase):port":
  this.sentinels = util._.uniq(this.sentinels, function (conf) {
    return conf.host.toLowerCase() + ":" + conf.port;
  });
};


RedisSentinel.prototype.kill = function kill() {
  if (this.terminated) { return; }
  this.terminated = true;

  this._log("Termination triggered from outside");

  // Kill our redis clients:
  if (this.fetcher) { this.fetcher.kill(); }
  if (this.watcher) { this.watcher.kill(); }
  this.fetcher = this.watcher = null;
};


// Exports:
module.exports.RedisSentinel = RedisSentinel;

// Internal API for passing along the client's Redis library:
module.exports.setClientRedis = function setClientRedis(mod) {
  client_redis = mod;
};
